lib/srv/git/forward.go

```diff
@@ -0,0 +1,590 @@
+		serverConn:       serverConn,
+		clientConn:       clientConn,
+		logger:           logger,
+		reply:            sshutils.NewReply(logger),
+		id:               uuid.NewString(),
+		waitExec:         make(chan error, 1),
+		verifyRemoteHost: verifyRemoteHost(cfg.TargetServer),
+		makeRemoteSigner: makeRemoteSigner,
+	}
+	// TODO(greedy52) extract common parts from srv.NewAuthHandlers like
+	// CreateIdentityContext and UserKeyAuth to a common package.
+	s.auth, err = srv.NewAuthHandlers(&srv.AuthHandlerConfig{
+		Server:       s,
+		Component:    teleport.ComponentForwardingGit,
+		Emitter:      s.cfg.Emitter,
+		AccessPoint:  cfg.AccessPoint,
+		TargetServer: cfg.TargetServer,
+		FIPS:         cfg.FIPS,
+		Clock:        cfg.Clock,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return s, nil
+
+}
+
+// Serve starts an SSH server that forwards git commands.
+func (s *ForwardServer) Serve() {
+	defer s.close()
+	s.logger.DebugContext(s.cfg.ParentContext, "Starting forwarding git")
+	defer s.logger.DebugContext(s.cfg.ParentContext, "Finished forwarding git")
+	server, err := sshutils.NewServer(
+		teleport.ComponentForwardingGit,
+		utils.NetAddr{}, /* empty addr, this is one time use so no use for listener*/
+		sshutils.NewChanHandlerFunc(s.onChannel),
+		sshutils.StaticHostSigners(s.cfg.HostCertificate),
+		sshutils.AuthMethods{
+			PublicKey: s.userKeyAuth,
+		},
+		sshutils.SetFIPS(s.cfg.FIPS),
+		sshutils.SetCiphers(s.cfg.Ciphers),
+		sshutils.SetKEXAlgorithms(s.cfg.KEXAlgorithms),
+		sshutils.SetMACAlgorithms(s.cfg.MACAlgorithms),
+		sshutils.SetClock(s.cfg.Clock),
+		sshutils.SetNewConnHandler(sshutils.NewConnHandlerFunc(s.onConnection)),
+	)
+	if err != nil {
+		s.logger.ErrorContext(s.cfg.ParentContext, "Failed to create git forward server", "error", err)
+		return
+	}
+	server.HandleConnection(s.serverConn)
+}
+
+func (s *ForwardServer) close() {
+	for _, closer := range []io.Closer{
+		s.serverConn,
+		s.clientConn,
+		s.cfg.TargetConn,
+	} {
+		if err := closer.Close(); err != nil && !utils.IsOKNetworkError(err) {
+			s.logger.WarnContext(s.cfg.ParentContext, "Failed to close", "error", err)
+		}
+	}
+}
+
+func (s *ForwardServer) userKeyAuth(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+	cert, ok := key.(*ssh.Certificate)
+	if !ok {
+		return nil, trace.BadParameter("unsupported key type")
+	}
+	if len(cert.Extensions[teleport.CertExtensionGitHubUserID]) == 0 {
+		return nil, trace.BadParameter("missing GitHub user ID")
+	}
+
+	// Verify incoming user is "git" and override it with any valid principle
+	// to bypass principle validation.
+	if conn.User() != gitUser {
+		return nil, trace.BadParameter("only git is expected as user for git connections")
+	}
+	if len(cert.ValidPrincipals) > 0 {
+		conn = sshutils.NewSSHConnMetadataWithUser(conn, cert.ValidPrincipals[0])
+	}
+
+	// Use auth.UserKeyAuth to verify user cert is signed by UserCA.
+	permissions, err := s.auth.UserKeyAuth(conn, key)
```

Comment:
This will also do regular RBAC on the `TargetServer` unless you also change the check for `h.c.Component == teleport.ComponentForwardingNode` in authhandlers.go

Reply:
there is a check to avoid that:
https://github.com/gravitational/teleport/blob/67c298bc115037aab83a7d501e87fce77a1924b9/lib/srv/authhandlers.go#L471-L473